android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt

```diff
@@ -0,0 +1,136 @@
+        try {
+            val serverRecords = RemoteSettings(config).use {
+                it.getRecords().records
+            }
+            return@withContext serverRecords
+        } catch (e: RemoteSettingsException.ResponseException) {
+            emptyList()
+        } catch (e: RemoteSettingsException) {
+            Logger.error(e.message.toString())
+            emptyList()
+
+        }
+    }
+
+    private fun writeRemoteRecordsToStorage(file: File, records: List<RemoteSettingsRecord>) {
+        try {
+            val jsonArray = JSONArray()
+            records.forEach {
+                val recordJson = JSONObject().apply {
+                    put("id", it.id)
+                    put("lastModified", it.lastModified.toString())
+                    put("deleted", it.deleted)
+                    put("attachment", it.attachment)
+                    put("fields", it.fields)
+                }
+                jsonArray.put(recordJson)
+            }
+
+            val jsonObject = JSONObject()
+            jsonObject.put("records", jsonArray)
+            AtomicFile(file).writeString { jsonObject.toString() }
+        } catch (e: Exception) {
+            Logger.error(e.message.toString())
+        }
+    }
+
+    /**
+     * Reads all records for a collection that can be found in the local storage.
+     **/
+    suspend fun readStorage(): List<RemoteSettingsRecord> = withContext(Dispatchers.IO) {
+        val folder = File(storageRootDirectory, bucketName)
+        val file = File(folder, collectionName)
+        var remoteRecords: List<RemoteSettingsRecord>
+        var records = mutableListOf<RemoteSettingsRecord>()
+
+        if (file.exists()) {
+            try {
+                val fileReader = FileReader(file)
+                val jsonObject = JSONObject(fileReader.readText())
+                val jsonArray = jsonObject.getJSONArray("records")
+                for (i in 0 until jsonArray.length()) {
+                    val recordJson = jsonArray.getJSONObject(i)
+                    val id = recordJson.getString("id")
+                    val lastModified = recordJson.getString("lastModified").toULong()
+                    val deleted = recordJson.getBoolean("deleted")
+                    val attachment = recordJson.opt("attachment")
+                    val fields = recordJson.opt("fields")
+
+                    records.add(
+                        RemoteSettingsRecord(
+                            id = id,
+                            lastModified = lastModified,
+                            deleted = deleted,
+                            attachment = attachment as Attachment?,
+                            fields = fields as RsJsonObject
+                        )
+                    )
+                    fileReader.close()
+
+                }
+                launch(Dispatchers.IO) {
+                    remoteRecords = fetchRemoteServerRecords()
```

Comment:
I think we should separate out fetching and writing from this function and rely on consumer to order their operations correctly. My main concern is that there is nothing in the name `readStorage` that suggests to me that this function will be initiating network IO and disk writes. 

Generally speaking, the advice that I try to follow for code (especially library code) is to try and avoid functions with the word "and" or "or" in them. This is the [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) in case you were not already aware of it. I think this function would need to be named something like `tryReadRecordsOrFetchAndWrite` to really capture all the behavior in it. 

This patch is most of the way there already anyway - I think we can just make `writeRemoteRecordsToStorage` public and do a few other tweaks. For example, we may want some kind of result type:

```kotlin
    sealed class RemoteSettingsResult {
        data class Success(val records: List<RemoteSettingsRecord>)
        data class DiskFailure(val error: Exception)
        data class NetworkFailure(val error: Exception)
    }
```

Reply:
I do think it would be fine to _also_ expose a method that tries to read from disk and then fetches from the network if it needs to, but I think we should provide more fine-grained control for consumers as a first step. An extension function would be a more idiomatic way to compose the two functions (and excuse a long name), IMO:

```kotlin
// top-level function
suspend fun RemoteSettingsClient.readOrFetch(): RemoteSettingsResult =
    this.read().takeIf { it is RemoteSettingsResult.Success } ?: this.fetch()
```

Reply:
I think the single responsibility principle makes very much sense and I sure will incorporate this in future too.